# Materailised Joins

## 0. Necessary imports and setups

In [57]:
from IPython.display import Image
from IPython.core.display import HTML
%load_ext sql
%sql postgresql://appdev@data:5432/appdev

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


'Connected: appdev@appdev'



## 1. On the table "circuits", report:

###          What type of indexes exists for the table and why they are of that type (and not some other type):

To find all the indexes for a specific table I use the following command in psql: "\d+ f1db.circuits" where we get the following:

In [16]:
%%html
<img src = "https://raw.githubusercontent.com/MartinH5/Materialised-Joins/master/Nummer1.PNG">

Here we can see 3 indexes: 

idx_17102_primary of the type btree,

idx_17102_url also with the type of btree

circuits_position_idx of the type gist.


bTree is a prefered structure for building indexing in database and is quite efficient at it. 
The gist, or Generalised Search Tree, is a versatile search structure, which allows us to build on whichever type we choose. Gist uses <key, pointer> pairs unlike the B-tree which uses integers.  

If we inspect each of the indexes we get the following sizes:

In [17]:
%%html
<img src = "https://raw.githubusercontent.com/MartinH5/Materialised-Joins/master/Nummer2.PNG">

## 2. Write a query that finds the winner of all the races, but only if they are younger than 38 years. The query should give return the date, driver surname, driver age, track time in milliseconds, race name and circuit name for all races.

In [67]:
%%sql
SELECT 
circuits.name AS Circuit_Name, 
races.name AS race_name, races.date AS race_date, 
drivers.surname AS driver_surname, drivers.dob AS driver_birthdate, 
results.milliseconds AS tracktime_ms FROM results
JOIN drivers ON drivers.driverid = results.driverid AND drivers.dob > '1980-03-19' AND results.position = 1
JOIN races ON races.raceid = results.raceid
JOIN circuits ON circuits.circuitid = races.circuitid
ORDER BY drivers.dob;

184 rows affected.


[('Istanbul Park', 'Turkish Grand Prix', datetime.date(2008, 5, 11), 'Massa', datetime.date(1981, 4, 25), 5209451),
 ('Autódromo José Carlos Pace', 'Brazilian Grand Prix', datetime.date(2006, 10, 22), 'Massa', datetime.date(1981, 4, 25), 5513751),
 ('Valencia Street Circuit', 'European Grand Prix', datetime.date(2008, 8, 24), 'Massa', datetime.date(1981, 4, 25), 5732339),
 ('Bahrain International Circuit', 'Bahrain Grand Prix', datetime.date(2007, 4, 15), 'Massa', datetime.date(1981, 4, 25), 5607515),
 ('Circuit de Spa-Francorchamps', 'Belgian Grand Prix', datetime.date(2008, 9, 7), 'Massa', datetime.date(1981, 4, 25), 4979394),
 ('Circuit de Barcelona-Catalunya', 'Spanish Grand Prix', datetime.date(2007, 5, 13), 'Massa', datetime.date(1981, 4, 25), 5496230),
 ('Istanbul Park', 'Turkish Grand Prix', datetime.date(2006, 8, 27), 'Massa', datetime.date(1981, 4, 25), 5331082),
 ('Bahrain International Circuit', 'Bahrain Grand Prix', datetime.date(2008, 4, 6), 'Massa', datetime.date(1981, 4, 25), 5466970),
 ('Circuit de Nevers Magny-Cours', 'French Grand Prix', datetime.date(2008, 6, 22), 'Massa', datetime.date(1981, 4, 25), 5510245),
 ('Autódromo José Carlos Pace', 'Brazilian Grand Prix', datetime.date(2008, 11, 2), 'Massa', datetime.date(1981, 4, 25), 5651435),
 ('Istanbul Park', 'Turkish Grand Prix', datetime.date(2007, 8, 26), 'Massa', datetime.date(1981, 4, 25), 5202161),
 ('Circuit de Monaco', 'Monaco Grand Prix', datetime.date(2006, 5, 28), 'Alonso', datetime.date(1981, 7, 29), 6223116),
 ('Silverstone Circuit', 'British Grand Prix', datetime.date(2006, 6, 11), 'Alonso', datetime.date(1981, 7, 29), 5151927),
 ('Circuit Gilles Villeneuve', 'Canadian Grand Prix', datetime.date(2006, 6, 25), 'Alonso', datetime.date(1981, 7, 29), 5677308),
 ('Silverstone Circuit', 'British Grand Prix', datetime.date(2011, 7, 10), 'Alonso', datetime.date(1981, 7, 29), 5321196),
 ('Suzuka Circuit', 'Japanese Grand Prix', datetime.date(2006, 10, 8), 'Alonso', datetime.date(1981, 7, 29), 5032413),
 ('Sepang International Circuit', 'Malaysian Grand Prix', datetime.date(2005, 3, 20), 'Alonso', datetime.date(1981, 7, 29), 5493736),
 ('Bahrain International Circuit', 'Bahrain Grand Prix', datetime.date(2005, 4, 3), 'Alonso', datetime.date(1981, 7, 29), 5358531),
 ('Autodromo Enzo e Dino Ferrari', 'San Marino Grand Prix', datetime.date(2005, 4, 24), 'Alonso', datetime.date(1981, 7, 29), 5261921),
 ('Nürburgring', 'European Grand Prix', datetime.date(2005, 5, 29), 'Alonso', datetime.date(1981, 7, 29), 5506648),
 ('Sepang International Circuit', 'Malaysian Grand Prix', datetime.date(2012, 3, 25), 'Alonso', datetime.date(1981, 7, 29), 9891812),
 ('Circuit de Nevers Magny-Cours', 'French Grand Prix', datetime.date(2005, 7, 3), 'Alonso', datetime.date(1981, 7, 29), 5482232),
 ('Hockenheimring', 'German Grand Prix', datetime.date(2005, 7, 24), 'Alonso', datetime.date(1981, 7, 29), 5188599),
 ('Valencia Street Circuit', 'European Grand Prix', datetime.date(2012, 6, 24), 'Alonso', datetime.date(1981, 7, 29), 6256649),
 ('Hockenheimring', 'German Grand Prix', datetime.date(2012, 7, 22), 'Alonso', datetime.date(1981, 7, 29), 5465862),
 ('Shanghai International Circuit', 'Chinese Grand Prix', datetime.date(2005, 10, 16), 'Alonso', datetime.date(1981, 7, 29), 5993618),
 ('Hungaroring', 'Hungarian Grand Prix', datetime.date(2003, 8, 24), 'Alonso', datetime.date(1981, 7, 29), 5941460),
 ('Marina Bay Street Circuit', 'Singapore Grand Prix', datetime.date(2008, 9, 28), 'Alonso', datetime.date(1981, 7, 29), 7036304),
 ('Fuji Speedway', 'Japanese Grand Prix', datetime.date(2008, 10, 12), 'Alonso', datetime.date(1981, 7, 29), 5421892),
 ('Sepang International Circuit', 'Malaysian Grand Prix', datetime.date(2007, 4, 8), 'Alonso', datetime.date(1981, 7, 29), 5534930),
 ('Circuit de Barcelona-Catalunya', 'Spanish Grand Prix', datetime.date(2013, 5, 12), 'Alonso', datetime.date(1981, 7, 29), 5956596),
 ('Circuit de Monaco', 'Monaco Grand Prix', datetime.date(2007, 5, 27), 'Alo

## 3. Describe the query using EXPLAIN ANALYZE with at least 5 lines of text. Answer at least the following

In [49]:
%%sql 
EXPLAIN ANALYSE
SELECT 
circuits.name AS Circuit_Name, 
races.name AS race_name, races.date AS race_date, 
drivers.surname AS driver_surname, drivers.dob AS driver_birthdate, 
results.milliseconds AS tracktime_ms FROM results
JOIN drivers ON drivers.driverid = results.driverid AND drivers.dob > '1980-03-19' AND results.position = 1
JOIN races ON races.raceid = results.raceid
JOIN circuits ON circuits.circuitid = races.circuitid
ORDER BY drivers.dob;

21 rows affected.


[('Sort  (cost=786.85..787.02 rows=68 width=62) (actual time=16.107..16.122 rows=184 loops=1)',),
 ('  Sort Key: drivers.dob',),
 ('  Sort Method: quicksort  Memory: 50kB',),
 ('  ->  Nested Loop  (cost=25.67..784.78 rows=68 width=62) (actual time=0.329..15.880 rows=184 loops=1)',),
 ('        ->  Nested Loop  (cost=25.53..772.81 rows=68 width=50) (actual time=0.316..15.485 rows=184 loops=1)',),
 ('              ->  Hash Join  (cost=25.25..746.50 rows=68 width=27) (actual time=0.277..10.842 rows=184 loops=1)',),
 ('                    Hash Cond: (results.driverid = drivers.driverid)',),
 ('                    ->  Seq Scan on results  (cost=0.00..708.96 rows=974 width=24) (actual time=0.036..10.367 rows=974 loops=1)',),
 ('                          Filter: ("position" = 1)',),
 ('                          Rows Removed by Filter: 22703',),
 ('                    ->  Hash  (cost=24.51..24.51 rows=59 width=19) (actual time=0.220..0.220 rows=60 loops=1)',),
 ('                          Buckets: 1024  Batches: 1  Memory Usage: 12kB',),
 ('                          ->  Seq Scan on drivers  (cost=0.00..24.51 rows=59 width=19) (actual time=0.009..0.197 rows=60 loops=1)',),
 ("                                Filter: (dob > '1980-03-19'::date)",),
 ('                                Rows Removed by Filter: 781',),
 ('              ->  Index Scan using idx_17199_primary on races  (cost=0.28..0.39 rows=1 width=39) (actual time=0.002..0.002 rows=1 loops=184)',),
 ('                    Index Cond: (raceid = results.raceid)',),
 ('        ->  Index Scan using idx_17102_primary on circuits  (cost=0.14..0.18 rows=1 width=28) (actual time=0.001..0.001 rows=1 loops=184)',),
 ('              Index Cond: (circuitid = races.circuitid)',),
 ('Planning time: 0.790 ms',),
 ('Execution time: 22.676 ms',)]

Running the Explain command plans how it will execute the queries. It will check the time it will take and if they is room for improvements. The Analyse command runs the querier and outputs the result above.

At the bottom we can see how long the planning took (0.79 ms) and how long it took to execute the query (22.676 ms). It explains all the operations the query had to go through to give us our desired result. It had to sort the table using quicksort. It had to loop through the tables. The tables had to be joined together. It had to consider indexing options for both races and circuits, using the primary indexing. We had to filter using their driver's birthdate and remove everyone above 38 years of age. There's a total of 10 calls. 

## 4. Create a materialized view of your query. Using EXPLAIN ANALYZE try to query the view. Write at least 5 lines of text explaining what's going on and why the query execution time changed

In [60]:
%%sql CREATE MATERIALIZED VIEW race_winners_cache AS
SELECT circuits.name AS Circuit_Name, 
races.name AS race_name, races.date AS race_date, 
drivers.surname AS driver_surname, drivers.dob AS driver_birthdate, 
results.milliseconds AS tracktime_ms FROM results
JOIN drivers ON drivers.driverid = results.driverid AND drivers.dob > '1980-03-19' AND results.position = 1
JOIN races ON races.raceid = results.raceid
JOIN circuits ON circuits.circuitid = races.circuitid
ORDER BY drivers.dob;

184 rows affected.


[]

In [66]:
%sql EXPLAIN ANALYZE SELECT * FROM race_winners_cache

3 rows affected.


[('Seq Scan on race_winners_cache  (cost=0.00..4.84 rows=184 width=65) (actual time=0.013..0.039 rows=184 loops=1)',),
 ('Planning time: 0.222 ms',),
 ('Execution time: 0.075 ms',)]

We can see the both planning and execution times are much faster than in the previous Explain Anaylse. Unlike the previous task, we only run a single Seq Scan now. This is because of the materialised view, which is a cached version of a normal view. We basically started off caching a view of the query and to get everything from this view, we just need to loop through it once, instead of having to create the view first. This make it much faster to analyse data. The disadvantage however, is the fact that it is a cached version of our data after all. If the original data where to get updated - we would still be operating on the old data.  